model/flow/header.go

```diff
@@ -80,7 +86,50 @@ func (h Header) ID() Identifier {
 	if h.Timestamp.Location() != time.UTC {
 		h.Timestamp = h.Timestamp.UTC()
 	}
-	return MakeID(h)
+
+	mutexHeader.Lock()
+
+	// unlock at the return
+	defer mutexHeader.Unlock()
```

Comment:
I don't quite understand this change. Isn't this lock would block two different threads computing IDs of two headers?

If engine A is calling header1.ID(), and engine B is calling header2.ID(), before this change, they can run concurrently, but with this change, one engine has to be blocked and wait until the other finishes its call. no?